src/java.base/share/classes/java/lang/runtime/ExactnessMethods.java

```diff
@@ -0,0 +1,281 @@
+
+    /** Exactness method from int to short
+     *
+     * @param n value
+     * @return  true if the passed value can be converted exactly to the target type
+     *
+     * */
+    public static boolean int_short(int n)     {return n == (int)(short)(n);}
+
+    /** Exactness method from int to char
+     *
+     * @param n value
+     * @return  true if the passed value can be converted exactly to the target type
+     *
+     * */
+    public static boolean int_char(int n)      {return n == (char)(n);}
+
+    /** Exactness method from int to float
+     *
+     * @param n value
+     * @return  true if the passed value can be converted exactly to the target type
+     *
+     * */
+    public static boolean int_float(int n) {
+        if (n == Integer.MIN_VALUE)
+            return true;
+        n = Math.abs(n);
+        return Float.PRECISION >= // 24
+                (32 - (Integer.numberOfLeadingZeros(n) +
+                        Integer.numberOfTrailingZeros(n))) ;
+    }
+
+    /** Exactness method from long to byte
+     *
+     * @param n value
+     * @return  true if the passed value can be converted exactly to the target type
+     *
+     * */
+    public static boolean long_byte(long n)    {return n == (long)(byte)(n);}
+
+    /** Exactness method from long to short
+     *
+     * @param n value
+     * @return  true if the passed value can be converted exactly to the target type
+     *
+     * */
+    public static boolean long_short(long n)   {return n == (long)(short)(n);}
+
+    /** Exactness method from long to char
+     *
+     * @param n value
+     * @return  true if the passed value can be converted exactly to the target type
+     *
+     * */
+    public static boolean long_char(long n)    {return n == (char)(n);}
+
+    /** Exactness method from long to int
+     *
+     * @param n value
+     * @return  true if the passed value can be converted exactly to the target type
+     *
+     * */
+    public static boolean long_int(long n)     {return n == (long)(int)(n);}
+
+    /** Exactness method from long to float
+     *
+     * @param n value
+     * @return  true if the passed value can be converted exactly to the target type
+     *
+     * */
+    public static boolean long_float(long n) {
+        if (n == Long.MIN_VALUE)
+            return true;
+        n = Math.abs(n);
+        return Float.PRECISION >= // 24
+                (64 - (Long.numberOfLeadingZeros(n) +
+                        Long.numberOfTrailingZeros(n))) ;
+    }
+
+    /** Exactness method from long to double
+     *
+     * @param n value
+     * @return  true if the passed value can be converted exactly to the target type
+     *
+     * */
+    public static boolean long_double(long n) {
+        if (n == Long.MIN_VALUE)
+            return true;
+        n = Math.abs(n);
+        return Double.PRECISION >= // 53
+                (64 - (Long.numberOfLeadingZeros(n) +
+                        Long.numberOfTrailingZeros(n))) ;
+    }
```

Comment:
Are you interested in optimizing the performance of these methods at this time?

For example, in this method:
* You can remove the `Long.MIN_VALUE` check.  That value "accidentally" works out with the rest of the logic; it has no leading zeros and 63 trailing zeros.  This optimization assumes that `Long.MIN_VALUE` isn't common enough to deserve its own fast path (at the expense of every other input), which seems true.
* If you assume that most inputs to this method are "small" integers that do fit in doubles (which I think is true), you can optimize for them by checking trailing zeros lazily.
  ```java
  n = Math.abs(n);
  int z = Long.numberOfLeadingZeros(n);
  return z > 10 || z + Long.numberOfTrailingZeros(n) > 10;
  ```

I can share benchmarks if that's useful.  (I needed this method years ago, and I wrote benchmarks back then.)

I'm assuming these methods are invoked whenever an expression like `longValue instanceof double` is evaluated, and that these methods aren't intrinsified (yet), so they seem worth optimizing.  But I could be mistaken.

Reply:
Thanks for leaving a comment @michaelhixson. The final implementation could contain more compact versions like `n == (long)(double)n && n != Long.MAX_VALUE;`. This prototype on the amber repo is to assess correctness for now. But indeed we can proceed by putting the less readable but more performant versions we have for some methods (@rgiulietti). You are right, these methods are not intrinsified.

Reply:
@michaelhixson The more compact variant given above in the comment by @biboudis, compiles to minimal machine code. There's hardly any need for an intrinsic.

Reply:
@rgiulietti In local benchmarks, the compact version of `long_double` is slower than what was in the PR originally, which in turn is slower than the "lazy trailing zeros" version in my previous comment.

Did you arrive at that compact form through benchmarking?  If so, great - my benchmarks might be bad or I might have conflicting results because I'm on Windows.  If not, I'll suggest that faster forms of this method may exist, which may be worth investigating... eventually.

Reply:
I added a benchmarks file (we may or may not keep it in the final PR) that we can discuss during the finalization of the PR on openjdk/jdk/. In any case, feel free to take a look on the test.

I followed these [instructions](https://openjdk.org/groups/build/doc/testing.html) for more info. 

The following are quick measurements on an Apple M1 Max chip/macOS 13.2.1 (22D68) -- quick meaning that if we want to keep them, we should run them with 3 forks for 15wi/15i at least.

```
sh make/devkit/createJMHBundle.sh

bash configure --with-boot-jdk="<.....>/prebuilt/jdk-19.jdk/Contents/Home/" --with-jmh=<.....>/jdk/make/devkit/../../build/jmh/jars/ -disable-warnings-as-errors

make test TEST="micro:org.openjdk.bench.jdk.preview.patterns.Exactness" MICRO="OPTIONS=-p "pollute=true";VM_OPTIONS=--enable-native-access=ALL-UNNAMED;RESULTS_FORMAT=json" CONF=rel

Exactness.test_float_int_based_on_compare             avgt    5  22001.714 ±  21.124  ms/op
Exactness.test_float_int_based_on_filtering           avgt    5  21459.505 ± 159.797  ms/op
Exactness.test_int_float_based_on_filtering           avgt    5   1464.611 ±   5.375  ms/op
Exactness.test_int_float_based_on_leading_trailing    avgt    5   2741.839 ±  10.300  ms/op
Exactness.test_long_double_based_on_filtering         avgt    5   1379.058 ±   4.329  ms/op
Exactness.test_long_double_based_on_leading_trailing  avgt    5   2439.624 ±   4.366  ms/op
Exactness.test_long_float_based_on_filtering          avgt    5   1491.983 ±   8.019  ms/op
Exactness.test_long_float_based_on_leading_trailing   avgt    5   2736.461 ±   9.508  ms/op
```

Interesting that both float_int are the same and I observe that the filtering ones (or compact) are better than all the analytical (leading_trailing).

Reply:
@biboudis It's interesting that your `long_double` results conflict with mine.

My results on an i7-7700K + Windows:
```
Benchmark                                             Mode  Cnt     Score     Error  Units
Exactness.test_long_double_based_on_filtering         avgt    5  4579.436 ±  66.288  ms/op
Exactness.test_long_double_based_on_leading_trailing  avgt    5  3322.384 ± 110.679  ms/op
```

My results on an i9-7960X + Windows:
```
Benchmark                                             Mode  Cnt     Score     Error  Units
Exactness.test_long_double_based_on_filtering         avgt    5  5403.940 ± 186.774  ms/op
Exactness.test_long_double_based_on_leading_trailing  avgt    5  3828.349 ±  81.673  ms/op
```

I've never been able to build the JDK, so I copied your test code into a standalone project with Java 19 and JMH 1.36.  So there's that, the processors, and the operating systems as possible sources of the difference.

I'm not sure how to tell whose results are more "correct", if there is such a thing.  I suppose that if the performance is a wash, the filtering approach is more attractive: it's easier to understand and it's stylistically similar to the rest of the exactness methods.

Thanks for trying out the benchmarks anyway.

Reply:
Seems like there are fast integer<->floating-point conversion instructions on Aarch64 with semantics very similar to Java's, but that on x86-64 the conversions are rather convoluted.

However, if the check is then followed by the cast, as in a successful `instanceof`, for example, then maybe the "filtering" variant works well even of x86-64, because the result might already be available in a register from the preceding check.

Hard to tell from the results above, though.